pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go

```diff
@@ -0,0 +1,248 @@
+// <name>Available: indicates that the CSI Controller Service was successfully deployed.
+// <name>Progressing: indicates that the CSI Controller Service is being deployed.
+// <name>Degraded: produced when the sync() method returns an error.
+type CSIDriverControllerServiceController struct {
+	name           string
+	manifest       []byte
+	operatorClient v1helpers.OperatorClient
+	kubeClient     kubernetes.Interface
+	deployInformer appsinformersv1.DeploymentInformer
+}
+
+func NewCSIDriverControllerServiceController(
+	name string,
+	manifest []byte,
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	deployInformer appsinformersv1.DeploymentInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &CSIDriverControllerServiceController{
+		name:           name,
+		manifest:       manifest,
+		operatorClient: operatorClient,
+		kubeClient:     kubeClient,
+		deployInformer: deployInformer,
+	}
+
+	return factory.New().WithInformers(
+		operatorClient.Informer(),
+		deployInformer.Informer(),
+	).WithSync(
+		c.sync,
+	).ResyncEvery(
+		time.Minute,
+	).WithSyncDegradedOnError(
+		operatorClient,
+	).ToController(
+		c.name,
+		recorder.WithComponentSuffix("csi-driver-controller-service_"+strings.ToLower(name)),
+	)
+}
+
+func (c *CSIDriverControllerServiceController) Name() string {
+	return c.name
+}
+
+func (c *CSIDriverControllerServiceController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	opSpec, opStatus, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if opSpec.ManagementState != opv1.Managed {
+		return nil
+	}
+
+	manifest := replacePlaceHolders(c.manifest, opSpec)
+	required := resourceread.ReadDeploymentV1OrDie(manifest)
+
+	deployment, _, err := resourceapply.ApplyDeployment(
+		c.kubeClient.AppsV1(),
+		syncContext.Recorder(),
+		required,
+		resourcemerge.ExpectedDeploymentGeneration(required, opStatus.Generations),
+	)
+	if err != nil {
+		return err
+	}
+
+	availableCondition := opv1.OperatorCondition{
+		Type:   c.name + opv1.OperatorStatusTypeAvailable,
+		Status: opv1.ConditionTrue,
+	}
+
+	if deployment.Status.AvailableReplicas > 0 {
+		availableCondition.Status = opv1.ConditionTrue
+	} else {
+		availableCondition.Status = opv1.ConditionFalse
+		availableCondition.Message = "Waiting for Deployment to deploy the CSI Controller Service"
+		availableCondition.Reason = "AsExpected"
+	}
+
+	progressingCondition := opv1.OperatorCondition{
+		Type:   c.name + opv1.OperatorStatusTypeProgressing,
+		Status: opv1.ConditionFalse,
+	}
+
+	if ok, msg := isProgressing(opStatus, deployment); ok {
+		progressingCondition.Status = opv1.ConditionTrue
+		progressingCondition.Message = msg
+		progressingCondition.Reason = "AsExpected"
+	}
+
+	updateStatusFn := func(newStatus *opv1.OperatorStatus) error {
+		// TODO: set ObservedGeneration (the last stable generation change we dealt with)
```

Comment:
Setting ObservedGeneration is IMO hard, as there are two independent controller that need to agree on its value.

Reply:
> Setting ObservedGeneration is IMO hard, as there are two independent controller that need to agree on its value.

This doesn't have a good solution in any controller yet.  Generically across other operators many different pieces need to agree.  I don't think this is significantly worse than other solutions we've attempted.